### PR TITLE
feat: support optimism

### PR DESCRIPTION
**Pull request by kien-rise**

### CODE REVIEW

File: crates/revm/src/optimism/handler_register.rs [hunk 1]
```diff
@@ -78,7 +87,7 @@ pub fn last_frame_return<SPEC: Spec, EXT, DB: Database>(
     let is_deposit = env.tx.optimism.source_hash.is_some();
     let tx_system = env.tx.optimism.is_system_transaction;
     let tx_gas_limit = env.tx.gas_limit;
-    let is_regolith = SPEC::enabled(REGOLITH);
+    let is_regolith = SPEC::enabled(SpecId::REGOLITH);
```
Comment (hai-rise) [inline crates/revm/src/optimism/handler_register.rs]: Avoid no-op changes to avoid potential conflicts on future rebases.

File: crates/revm/src/handler.rs [hunk 2]
```diff
@@ -77,19 +77,19 @@ impl<'a, EXT, DB: Database> EvmHandler<'a, EXT, DB> {
 
     /// Handler for optimism
     #[cfg(feature = "optimism")]
-    pub fn optimism<SPEC: Spec>() -> Self {
-        let mut handler = Self::mainnet::<SPEC>();
+    pub fn optimism<SPEC: Spec>(with_reward_beneficiary: bool) -> Self {
+        let mut handler = Self::mainnet::<SPEC>(with_reward_beneficiary);
         handler.cfg.is_optimism = true;
-        handler.append_handler_register(HandleRegisters::Plain(
```
Comment (hai-rise) [inline crates/revm/src/handler.rs:83]: Why the change in signature and going from `Plain` to `Box`?

File: crates/revm/src/optimism/handler_register.rs [hunk 3]
```diff
@@ -18,25 +18,34 @@ use revm_precompile::{secp256r1, PrecompileSpecId};
...
-        handler.post_execution.output = Arc::new(output::<SPEC, EXT, DB>);
-        handler.post_execution.end = Arc::new(end::<SPEC, EXT, DB>);
-    });
+pub fn optimism_handle_register<DB: Database, EXT>(
+    with_reward_beneficiary: bool,
+) -> HandleRegisters<EXT, DB> {
+    HandleRegisters::Box(Box::new(move |handler: &mut EvmHandler<'_, EXT, DB>| {
+        spec_to_generic!(handler.cfg.spec_id, {
+            // validate environment
+            handler.validation.env = Arc::new(validate_env::<SPEC, DB>);
```
Comment (hai-rise) [inline crates/revm/src/optimism/handler_register.rs:28]: For performance we should downgrade these `Arc`s into `Box`es as well.

**Comment by kien-rise** [comment]
moved to https://github.com/risechain/revm/pull/2

File: crates/revm/src/handler.rs (hunk 2 above)
Comment (kien-rise) [reply crates/revm/src/handler.rs:83]: I reverted the signature change in https://github.com/risechain/revm/pull/2

File: crates/revm/src/optimism/handler_register.rs (hunk 3 above)
Comment (kien-rise) [reply crates/revm/src/optimism/handler_register.rs:28]: this should be done on `main` branch instead, right?

Comment (kien-rise) [reply crates/revm/src/optimism/handler_register.rs:28]: let's move this discussion to #2 if needed
